### DISCUSSION TITLE
qqplot for tweedie model

### DISCUSSION THREAD

**Discussion by IosuParadinas** [inlabru-org/inlabru#211] [Q&A]
Hi!

I am trying to create a qqplot for a tweedie model. I looked into the _mgcv_ package to see how they do it and I believe that I am stuck in the calculation of deviance residuals for this tweedie distribution. I need generate smaples from the model and manually create deviance residuals using this samples. Inlabru provides residuals in the output, but I am being incapable to replicate them by hand. I tried to dig in the _bru()_ function to find where/how the residuals are computed but I havent been able to find it.

Could anybody guide me on where I could find this?

Here´s a example code. I would need to set the _deviance.residuals()_ function below to get the qqplot. Many thnaks!

**data and model fit**
_library(inlabru)
library(tweedie)

n <-100
x <- rnorm(n, mean = 0, sd = 0.2)
eta <- 1 + .5*x #- 7.1*x^2
mu <- exp(eta)
p <- 1.1
phi <- 1.75
y <- numeric(n)
for(i in 1:n) {
  y[i] <- rtweedie(1, xi = p, mu = mu[i], phi = phi)
  }
r <- bru(y ~ 1 + f(x,model="linear"), data = data.frame(y, x), 
family = "tweedie")_

**1- simulate model residuals by generating new data based on the fitted model and calculate the residuals for that new data**

_for(j in 1:rep){
 sample_model=c()
  for(i in 1:n) {
    sample_model [i] <- rtweedie(1, xi = r$summary.hyperpar[1,1], mu = fitted_mu$mean[i], phi = r$summary.hyperpar[2,1])
  }
dm[,j] = deviance.residuals(y,sample_model) ## simulate residuals
}_

**2- get the empirical quantiles of the simulated residuals obtained**

_Dq = quantile(as.numeric(dm), (1:n - 0.5)/n)_

**3- Do the qqplot**

_qqplot(Dq,D)
abline(0,1)_

**Comment by finnlindgren** [comment]
Hi, residuals in the Bayesian setting can have more than one meaning; e.g. is it the posterior expectation of a residual, or the residual of the posterior expectation, or (as it seems you want to do) the posterior distribution of residuals; the expectations are the same when the residual is a simple difference between observation and linear predictor, but is not the same for non-linear residual definitions, which I think includes deviance residuals, and the "posterior distribution of residuals" is a separate thing as well. Do you have the specific mathematical definition of deviance residual for your special case?  It loos like from your code step "1" that you're using the approach of simulating posterior predictive samples from the model, and that the deviance.residual is somehow determined with the help of that?

**Reply by finnlindgren** [reply]
There may be some useful information in https://inlabru-org.github.io/inlabru/articles/prediction_scores.html and/or https://inlabru-org.github.io/inlabru/articles/2d_lgcp_residuals.html (the latter is for point processes, but some of the techniques may be relevant).

**Reply by finnlindgren** [reply]
I found this paper that contains an actual definition of deviance residuals (was surprisingly difficult to find an actual definition!): https://doi.org/10.1186/s12874-020-01055-2
There, they define it as the signed square root of the difference in log-densities between the estimated model and the "saturated model". As finding what that expression is needs to be done for any given distribution (and also need to decide how to deal with the Bayesian posterior distribution aspect; the residuals in the paper are defined for the maximum likelihood estimate, not in a Bayesian context) you likely need to find a paper explicitly discussing how to compute deviance residuals for the Tweedie distribution to find out what actually needs to be computed.

**Reply by IosuParadinas** [reply]
Thanks Finn! Sorry for my late reply, I´ve been away.

Yes, I´ve been unable to find a definition for the tweedie. And I had not thought about the posterior of residuals. I was thining to use a simple difference between observation and linear predictor, but you are right.

Said this, the inlabru model object provides deviance residuals so I thought the software already had this expression built in when choosing one distribution or the other.

**Reply by finnlindgren** [reply]
Yes, it's quite possible that it already calculates deviance residuals.
